File: solutions/go/sieve/sieve.go
package sieve

// Sieve of Eratosthenes algorithm for finding all prime numbers up to the given limit.
func Sieve(limit int) (primes []int) {
	if limit < 2 {
		return
	}

	list := make([]bool, limit+1)
	list[0] = true
	list[1] = true

	for i, marked := range list {
		if !marked {
			for j := i * i; j <= limit; j += i {
				list[j] = true
			}
			primes = append(primes, i)
		}
	}
	return
}

File: solutions/cpp/sieve/sieve.cpp
#include "sieve.h"

// Sieve of Eratosthenes algorithm for finding all prime numbers up to the given limit.
namespace sieve
{
    using namespace std;

    vector<int> primes(int limit)
    {
        if (limit < 2)
        {
            return vector<int>{};
        }

        vector<int> primes{2};
        vector<bool> marked(limit + 1);

        for (int i = 3; i <= limit; i += 2)
        {
            if (marked[i])
            {
                continue;
            }
            for (int j = i * i; j <= limit; j += i)
            {
                marked[j] = true;
            }
            primes.emplace_back(i);
        }
        return primes;
    }
} // namespace sieve
